staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go

```diff
@@ -237,11 +249,49 @@ func (f *sharedInformerFactory) InformerFor(obj {{.runtimeObject|raw}}, newFunc
 var sharedInformerFactoryInterface = `
 // SharedInformerFactory provides shared informers for resources in all known
 // API group versions.
+//
+// It is typically used like this:
+//
+//   ctx, cancel := context.Background()
+//   defer cancel()
+//   factory := NewSharedInformerFactory(client, resyncPeriod)
+//   defer factory.WaitForShutdown()    // Returns immediately if nothing was started.
+//   genericInformer := factory.ForResource(resource)
+//   typedInformer := factory.SomeAPIGroup().V1().SomeType()
+//   factory.Start(ctx.Done())          // Start processing these informers.
+//   synced := factory.WaitForCacheSync(ctx.Done())
+//   for v, ok := range synced {
+//       if !ok {
+//           fmt.Fprintf(os.Stderr, "caches failed to sync: %v", v)
+//           return
+//       }
+//   }
+//
+//   // Creating informers can also be created after Start, but then
+//   // Start must be called again:
+//   anotherGenericInformer := factory.ForResource(resource)
+//   factory.Start(ctx.Done())
 type SharedInformerFactory interface {
 	{{.informerFactoryInterface|raw}}
-	ForResource(resource {{.schemaGroupVersionResource|raw}}) (GenericInformer, error)
+
+	// Start initializes all requested informers. They are handled in goroutines
+	// which run until the stop channel gets closed.
+        Start(stopCh <-chan struct{})
+
+	// WaitForShutdown blocks until all goroutines have stopped.
```

Comment:
I'm not sure this doc-comment gives enough attention to the case we discussed earlier in this thread: https://github.com/kubernetes/kubernetes/pull/112200#discussion_r961758712

Reply:
There I suggested "WaitForShutdown waits for the factory to have no goroutines running anymore." Isn't "until all goroutines have stopped" saying the same, just with fewer words?

Reply:
I suppose. Personally I read the two differently:

"all goroutines have stopped" - does this refer to all goroutines at start of function call or at the time of shutdown? My initial read is start of function call.

"no goroutines are running anymore" - this communicates more directly to me that at the time of stopping, there will be no goroutines running. But that is just my read of it. You are right others might not find that sufficient also.

Maybe that is good reason enough to expound on it in a few more words. To me it is an important detail to get across: if different stop channels are used for the multiple calls to `Start`, the semantics for this function change.

Reply:
> "all goroutines have stopped" - does this refer to all goroutines at start of function call or at the time of shutdown?

There is no "time of shutdown". When `WaitForShutdown` is called, it checks for running goroutines and blocks until those terminate.

> if different stop channels are used for the multiple calls to Start, the semantics for this function change.

I don't see the difference. "all goroutines" includes "all" goroutines, regardless how they were started. So does "no goroutines are running".

Let's hear from someone else. I can change the wording, I just don't see why and would like to understand that first.

Reply:
> There is no "time of shutdown". When WaitForShutdown is called, it checks for running goroutines and blocks until those terminate.

Yes, and if any new goroutines are added while blocking, `WaitForShutdown` will wait for those too. This is a huge difference from just waiting for routines that existed before blocking. The current doc-comment does not communicate this nuance.

> I don't see the difference. "all goroutines" includes "all" goroutines, regardless how they were started. So does "no goroutines are running".

Sure, then perhaps something different from those wordings.

> Let's hear from someone else. I can change the wording, I just don't see why and would like to understand that first.

👍🏼

Reply:
The question I'd answer in the comment is whether new goroutines can start or them all having stopped is a permanent condition. "Shutdown" implies that it's permanent, but both proposed words don't demand this interpretation.

Reply:
So the problem is not the function *comment*, but the function *name*. How about this (pushed):

	// WaitForStop blocks until all goroutines have terminated. For that
	// the close channel(s) that they were started with must be closed,
	// either before WaitForStop gets called or while it is waiting.
	// It is possible to start and stop multiple times.
	WaitForStop()

Reply:
Yeah, that's better! It raises some followup questions, mostly do we really recommend that people call this multiple times and/or with different stop channels?

From an implementation point of view, I think it may be impossible to use this safely (if there are multiple starts that the author of the call to WaitForStop() is unaware of) since the code is using a WaitGroup and: https://pkg.go.dev/sync#WaitGroup.Add

Reply:
> mostly do we really recommend that people call this multiple times and/or with different stop channels?

The existing API allows that, so in that sense I am only documenting what people may have already started to rely upon. I don't think we should disallow it now.

> it may be impossible to use this safely

Hmm, what can go wrong?

With "multiple times" I meant:
```
factory.Start(...)
# close channel
factory.WaitForStop()

factory.Start(...)
# close channel
factory.WaitForStop()
```

That is safe. What is non-deterministic is calling Start and WaitForStop in different goroutines without coordination. The WaitForStop may have already returned when Start increments the counter again.

The WaitGroup docs says: "Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.".

I think this just means that `wg.Wait()` followed by `wg.Add(1)` will not work because `wg.Wait()` returns immediately. But that's consistent with how `WaitForStop` is defined: if it gets called before `Start`, then no goroutines were started and thus (logically true) "all goroutines (that were started so far) have terminated".

Reply:
Right. If the caller doesn't have any synchronization, they could call WaitForStop() and as of that moment everything is stopped, but by the time they do their next line (whatever that is) a new goroutine could have started again. Functions like this usually don't require synchronization like that because usually being stopped is a permanent condition. If it's not permanent, it's potentially hard to use this safely, depending on what the caller is going to do next.

Why are people going to call WaitForStop? The only reason that comes to mind is that we're going to shut down completely and don't want undrained queues or extra goroutines.

My first thought on how to make this safe is to just make it permanent:
* Change the name to just Stop() or Shutdown()
* When called, set a "shuttingDown" flag to true
* Make start/run refuse to do so if the flag is true
* Document that caller must also close all stop channels (alternatively, you could store them and close them for the caller, but that seems like maybe too much work)

This won't harm existing callers because there's no shutdown method at all. Your intended use case of test cases will never need to re-start a stopped thingy.

Reply:
The use case is indeed mostly unit tests.

A `Shutdown` function with a behavior as described works for me, there's just one problem: should `Start` then be changed to return an error (currently it has none) or just silently not start informers?

I suppose it has to be the latter, because adding an error would force everyone to add error checking to avoid linter errors for a situation that never occurs in their code.

Reply:
Yeah, that's a dilemma...

One more option is to panic in that condition.

I guess yeah, make the existing `Start` silently ignore this. Add on to the comment...

I don't feel great about that because it'll be hard to notice if it does happen to someone, but the other options seem to involve more work for more people.

Reply:
I pushed a version with `Shutdown`.